### PR TITLE
Job scheduling

### PR DESCRIPTION
**Pull request by mitchelljy**
Here's an initial PR for [T19](https://projects.backbeat.tech/projects/1031eb11-10e9-11e8-a781-f23c916e1390/tasks/19)

I decided to go with this cron package: https://godoc.org/github.com/robfig/cron

It handles expressions _and_ scheduling them. Is that what we want? In the task discussion you mentioned that the agent could check every x minutes and decide which backups it should run which is more manual, but this package will handle all of that for us.

For now the main loop runs every minute, and it will randomly assign each backup it pulls a cron expression for run every 1-2 minutes.

### CODE REVIEW

File: command/agent.go [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
+import (
+	"log"
+	"net/http"
+	"time"
+
+	"../actions"
+	"../auth"
+	"../config"
+	"../utils"
+	"github.com/robfig/cron"
```
Comment (glynnforrest) [inline command/agent.go:12]: Can you update the `Makefile` to make sure this dependency is included?

File: command/agent.go [hunk 2]
```diff
@@ -25,32 +29,29 @@ var Agent = cli.Command{
...
+		crons := map[string]*cron.Cron{}
 
-		body, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			log.Fatal("Error reading body. ", err)
-		}
+		cr := cron.New()
+		cr.AddFunc("* * * * *", func() {
+			log.Println("Checking for changes to backups...")
+			newBackups := actions.ListBackups(client, tokenResponse, actions.BACKUP_TYPE_SCHEDULED)
```
Comment (glynnforrest) [inline command/agent.go]: Here would be a good spot for some more logging. Seeing "Checking for changes to backups..." in the output with no other lines could be disconcerting. I'd like to see the number of backups fetched, even if none were found or none of them are scheduled. Even a debug log line showing "Found 5 backups, all are already scheduled" would be very useful.

File: command/agent.go [hunk 3]
```diff
@@ -25,32 +29,29 @@ var Agent = cli.Command{
...
-		}
-		auth.AddAuthHeader(req, tokenResponse)
 
-		resp, err := client.Do(req)
-		if err != nil {
-			log.Fatal("Error reading response. ", err)
-		}
-		defer resp.Body.Close()
+		backups := map[string]actions.Backup{}
+		crons := map[string]*cron.Cron{}
```
Comment (glynnforrest) [inline command/agent.go]: This is an elegant solution, very nice!

File: command/agent.go [hunk 4]
```diff
@@ -25,32 +29,29 @@ var Agent = cli.Command{
...
 
-		body, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			log.Fatal("Error reading body. ", err)
-		}
+		cr := cron.New()
+		cr.AddFunc("* * * * *", func() {
```
Comment (glynnforrest) [inline command/agent.go]: Can we run this function immediately when the agent starts, as well as every minute? I'd probably extract the anonymous function to a named function, then call it.

```go
check() // pick a better name than me
cr.AddFunc("* * * * *", check)
```

File: utils/scheduler.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"fmt"
+	"log"
+	"net/http"
+
+	"../actions"
+	"../auth"
+	"github.com/robfig/cron"
+)
+
+func Schedule(client *http.Client, tokenResponse auth.AccessTokenResponse, backup actions.Backup) *cron.Cron {
```
Comment (glynnforrest) [inline utils/scheduler.go:13]: You've probably noticed a pattern emerging here, we'll refactor to use an ApiClient struct instead. Then we'd pass it into this function: `Schedule(apiClient, backup)`

**Review by glynnforrest — CHANGES_REQUESTED** [review]
Really great work here! I like how you're scheduling the backups. Nice find with the cron package too.

A few changes please:

* More logging (no such thing as too much, check out what consul or the salt-minion dump to their logs)
* Run the backup check immediately on startup as well as every minute
* Cron package in Makefile


We're at the point now where passing a base `*http.Client` and `tokenResponse` around is a bit repetitive, we should probably create an `ApiClient` struct with methods on it. Let's cover that in a pairing session together 👍

**Comment by mitchelljy** [comment]
@glynnforrest That's all of those changes pushed. Is that a suitable amount of logging for the main loop?

Other changes: 
* Updated to use cron expressions from the API as per your branch.
* The sync frequency (not sure about that terminology, what do you think?) can now be specified as a CLI argument.

**Review by glynnforrest — APPROVED** [review]

**Pull request merged**